source op support s and fixed generator bug

grybd opened the pull request:
#### 这个PR的目的
- [x] randperm op支持 S0、添加单测
- [x]  处理 random op 支持s 过程中，各个rank间local tensor 总是相等的bug 相关背景记录在https://github.com/Oneflow-Inc/oneflow/pull/7434#issuecomment-1033306931
####  random op 支持 global tensor 一致性
1) 在处理 randint op 和  rand op 支持B/S 保持global tensor 的一致性所采取的方案是利用 GetOpKernelRandomSeed(ctx)这个工具函数进行设计，当op 支持 S时 不同rank 间调用GetOpKernelRandomSeed(ctx) 返回一个不同的seed，再通过generator->set_current_seed(ctx->Attr<int64_t>("seed") + GetOpKernelRandomSeed(ctx)) 就可以为每个rank 设计不同的seed,这样能保证uniform 类的kernel 经过S 生成 同分布 不同数值的local tensor ，当op支持B时 每个rank 上 kernel 调用GetOpKernelRandomSeed(ctx) 时会生成相同的seed ,再通过generator->set_current_seed(ctx->Attr<int64_t>("seed") + GetOpKernelRandomSeed(ctx))  就保证了每个rank 都拿到了相同的seed,这样就可以保持global tensor 的一致性
2) 在处理 randperm op 和 arange op 支持 B/S 时 保持 global tensor 的一致性，目前打算采用的处理方案是让多个rank 公用seed 然后在 先在每个rank上生成完整的tensor再根据 infer physic shape信息利GetTensorSliceView4ParallelId(parallel_hierarchy, nd_sbp, logical_shape, parallel_id) 这个工具函数，获得本rank_id 和  physic shape所对应的tensor 上的索引信息，再把对应的位置的数据拷贝到 本rank 的local tensor 上 

以上方案 是通过与xiaoyu，yinggang开会总结出来的


fixed: https://github.com/Oneflow-Inc/OneTeam/issues/1167

strint commented on oneflow/user/kernels/distributions/uniform_int_kernel.h:
```diff
@@ -75,7 +76,8 @@ class UniformIntKernel final : public user_op::OpKernel {
     const auto& generator = CHECK_JUST(one::MakeAutoGenerator());
-    generator->set_current_seed(ctx->Attr<int64_t>("seed"));
+    int64_t rank_id = GlobalProcessCtx::Rank();
+    generator->set_current_seed(ctx->Attr<int64_t>("seed")+rank_id);
```
需要format一下

strint replied:
需要注释一下

grybd replied:
已经处理~

strint commented on oneflow/user/kernels/distributions/uniform_kernel.h:
```diff
@@ -33,7 +34,8 @@ class UniformKernel final : public user_op::OpKernel {
     const auto& generator = CHECK_JUST(one::MakeGenerator(device_type));
-    generator->set_current_seed(ctx->Attr<int64_t>("seed"));
+    int64_t rank_id = GlobalProcessCtx::Rank();
+    generator->set_current_seed(ctx->Attr<int64_t>("seed")+rank_id);
```
同上

grybd replied:
ok

strint commented on oneflow/user/kernels/randperm_kernel.cpp:
```diff
@@ -29,7 +31,8 @@ class CpuRandPermKernel final : public user_op::OpKernel {
     const auto& generator = CHECK_JUST(one::MakeGenerator(kCPU));
-    generator->set_current_seed(ctx->Attr<int64_t>("seed"));
+    int64_t rank_id = GlobalProcessCtx::Rank();
+    generator->set_current_seed(ctx->Attr<int64_t>("seed")+rank_id);
```
同上

grybd replied:
ok

strint commented on oneflow/user/kernels/randperm_kernel.cu:
```diff
@@ -42,7 +43,8 @@ class GpuRandPermKernel final : public user_op::OpKernel {
     const auto& generator = CHECK_JUST(one::MakeGenerator(kCUDA));
-    generator->set_current_seed(ctx->Attr<int64_t>("seed"));
+     int64_t rank_id = GlobalProcessCtx::Rank();
+    generator->set_current_seed(ctx->Attr<int64_t>("seed")+rank_id);
```
同上

grybd replied:
ok

strint commented on oneflow/user/ops/randperm_op.cpp:
```diff
@@ -32,7 +33,22 @@ namespace oneflow {
+  DimVector dim_vec{n};
+  const Shape& shape = Shape(dim_vec);
+  const cfg::SbpParallel& out_sbp_para = ctx->SbpParallel4ArgNameAndIndex("out", 0);
+  if (out_sbp_para.has_split_parallel()) {
```
这里只处理了1D的。我们没问题了。

辛苦 @wyg1997 指导下，把2D的支持 + 测试也加上。

strint commented on oneflow/user/ops/randperm_op.cpp:
```diff
@@ -32,7 +33,22 @@ namespace oneflow {
+  DimVector dim_vec{n};
+  const Shape& shape = Shape(dim_vec);
+  const cfg::SbpParallel& out_sbp_para = ctx->SbpParallel4ArgNameAndIndex("out", 0);
+  if (out_sbp_para.has_split_parallel()) {
```
这里只支持了1D的，辛苦 @wyg1997 review指导下，再加上2D的支持和测试。

wyg1997 replied:
这里照搬 https://github.com/Oneflow-Inc/oneflow/pull/7556/files#diff-0d787cb4011909380ae7ee4369b05972fc72c6a9beb3ebf94d26d8aa93a3f8b6R156 这里的逻辑就可以，总体逻辑就是构造出 GetTensorSliceView4ParallelId 函数所需要的参数，然后直接拿结果赋值就可以了。

grybd replied:
已经参考改了下，麻烦有空帮忙review 一下~ @strint @wyg1997

wyg1997 commented on oneflow/user/ops/randperm_op.cpp:
```diff
@@ -32,8 +34,20 @@ namespace oneflow {
+  const Shape& parallel_hierarchy = *ctx->parallel_desc().hierarchy();
+  const NdSbp& nd_sbp = ctx->NdSbp4ArgNameAndIndex("out", 0);
+  int32_t n = ctx->Attr<int32_t>("n");
+  DimVector dim_vec{n};
+  const Shape& logical_shape = Shape(dim_vec);
```
```suggestion
  const Shape& logical_shape = Shape({n});
```

grybd replied:
> 

ok

wyg1997 commented on oneflow/user/kernels/distributions/uniform_int_kernel.h:
```diff
@@ -75,7 +76,11 @@ class UniformIntKernel final : public user_op::OpKernel {
   std::shared_ptr<user_op::OpKernelState> CreateOpKernelState(
       user_op::KernelInitContext* ctx) const override {
     const auto& generator = CHECK_JUST(one::MakeAutoGenerator());
-    generator->set_current_seed(ctx->Attr<int64_t>("seed"));
+    int64_t rank_id = GlobalProcessCtx::Rank();
+    generator->set_current_seed(
+        ctx->Attr<int64_t>("seed")
+        + rank_id);  // each rank get a different seed to generate the same distribution, \
```
如果是 broadcast 的话，各个 rank 的数据能一致吗？

wyg1997 replied:
还有这里注释的格式比较奇怪，提到前面吧

strint replied:
eager是不是也不能一致？

wyg1997 replied:
是，eager 要传相同的 generator 才行

grybd replied:
哦哦，感觉确实是broadcast 的话，各个 rank 的数据好像是不能保证一致性，那这里还是得提前做一下判断如果是S才采取这种方案

wyg1997 replied:
讨论的方案可以稍微在 PR 里描述一下

grybd replied:
@wyg1997  好的，明天补充一下~

strint commented on oneflow/user/kernels/distributions/uniform_int_kernel.h:
```diff
@@ -75,7 +76,11 @@ class UniformIntKernel final : public user_op::OpKernel {
+    generator->set_current_seed(
+        ctx->Attr<int64_t>("seed")
+        + rank_id);  // each rank get a different seed to generate the same distribution, \
+                                                                       but different values of local tensor
```
注释写到79行前面吧

grybd replied:
好的

strint commented on python/oneflow/test/modules/test_arange.py:
```diff
@@ -95,5 +95,44 @@ def test_global_naive(test_case):
+            flow.placement("cpu", [0, 1]),
+            flow.placement("cuda", [0, 1]),
+        ]
+        arg_dict["sbp"] = [(flow.sbp.broadcast,), (flow.sbp.split(0),)]
```
这里增加下2D的测试？

grybd replied:
好的

grybd commented:
testcase 还没跑通, 我明天再看看~，麻烦有空也帮忙review一下哈 @strint @wyg1997

grybd commented:
跑./test_consistent_arange.py会报错，报错信息：
 File "/home/fengdaochao/fdz/oneflow/oneflow/core/job/job_build_and_infer_ctx.cpp", line 630, in AddAndInferOp
    CheckOpBlobSplitability(op, parallel_desc.parallel_num())
  File "/home/fengdaochao/fdz/oneflow/oneflow/core/job/job_build_and_infer_ctx.cpp", line 335, in CheckOpBlobSplitability
    Check failed: (current_shape.At(axis) % parallel_hierarchy->At(i)) == (0) (1 vs 0) op_name: arange-0 blob_name: out_0 cannot split blob by parallel_hierarchy: 2

跑./test_consistent_randperm.py会报错，报错信息：
F20220224 15:16:32.818838 986938 user_kernel.cpp:171] Check failed: parallel_desc_.hierarchy()->NumAxes() == 1 (2 vs. 1) 
*** Check failure stack trace: ***
F20220224 15:16:32.819939 986955 user_kernel.cpp:171] Check failed: parallel_desc_.hierarchy()->NumAxes() == 1 (2 vs. 1) 
*** Check failure stack trace: ***
    @     0x7f1cd9c17f97  google::LogMessage::Fail()
    @     0x7f1266515f97  google::LogMessage::Fail()
    @     0x7f1cd9c17ed8  google::LogMessage::SendToLog()
    @     0x7f1266515ed8  google::LogMessage::SendToLog()
    @     0x7f1cd9c177bb  google::LogMessage::Flush()
    @     0x7f12665157bb  google::LogMessage::Flush()
    @     0x7f1cd9c1b31c  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f126651931c  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f1cd3066957  oneflow::UserKernelInitAndCacheContext::SbpParallel4ArgNameAndIndex()
    @     0x7f125f964957  oneflow::UserKernelInitAndCacheContext::SbpParallel4ArgNameAndIndex()
    @     0x7f1cd5c83fd5  oneflow::GetOpKernelSeed()
    @     0x7f1cd5c85e81  oneflow::CpuRandPermKernel::CreateOpKernelState()
    @     0x7f1262581fd5  oneflow::GetOpKernelSeed()
    @     0x7f1262583e81  oneflow::CpuRandPermKernel::CreateOpKernelState()
    @     0x7f1cd3060655  oneflow::UserKernel::CreateOpKernelState()
    @     0x7f1cd3060ad8  oneflow::UserKernel::VirtualKernelInit()
    @     0x7f125f95e655  oneflow::UserKernel::CreateOpKernelState()
    @     0x7f1cd2fea86d  oneflow::Kernel::Init()
    @     0x7f125f95ead8  oneflow::UserKernel::VirtualKernelInit()
    @     0x7f1cd2feacef  oneflow::ConstructKernel()
    @     0x7f1cd30a6e95  oneflow::Actor::Init()
    @     0x7f125f8e886d  oneflow::Kernel::Init()
    @     0x7f1cd30bdfa9  oneflow::NewActor()
    @     0x7f125f8e8cef  oneflow::ConstructKernel()
    @     0x7f1cd44799c9  oneflow::Thread::ConstructActor()
    @     0x7f1cd4479411  oneflow::Thread::PollMsgChannel()
    @     0x7f125f9a4e95  oneflow::Actor::Init()
    @     0x7f1cd4478ae7  _ZZN7oneflow6ThreadC4ERKNS_8StreamIdEENKUlvE_clEv
    @     0x7f1cd447a4e3  _ZSt13__invoke_implIvZN7oneflow6ThreadC4ERKNS0_8StreamIdEEUlvE_JEET_St14__invoke_otherOT0_DpOT1_
    @     0x7f1cd447a484  _ZSt8__invokeIZN7oneflow6ThreadC4ERKNS0_8StreamIdEEUlvE_JEENSt15__invoke_resultIT_JDpT0_EE4typeEOS7_DpOS8_
    @     0x7f125f9bbfa9  oneflow::NewActor()
    @     0x7f1cd447a422  _ZNSt6thread8_InvokerISt5tupleIJZN7oneflow6ThreadC4ERKNS2_8StreamIdEEUlvE_EEE9_M_invokeIJLm0EEEEvSt12_Index_tupleIJXspT_EEE
    @     0x7f1cd447a3e3  _ZNSt6thread8_InvokerISt5tupleIJZN7oneflow6ThreadC4ERKNS2_8StreamIdEEUlvE_EEEclEv
    @     0x7f1cd447a3b8  _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN7oneflow6ThreadC4ERKNS3_8StreamIdEEUlvE_EEEEE6_M_runEv
    @     0x7f1c7f0d8de4  (unknown)
    @     0x7f1d2c76e609  start_thread
    @     0x7f1d2c8aa293  clone
    @     0x7f1260d779c9  oneflow::Thread::ConstructActor()
    @     0x7f1260d77411  oneflow::Thread::PollMsgChannel()
    @     0x7f1260d76ae7  _ZZN7oneflow6ThreadC4ERKNS_8StreamIdEENKUlvE_clEv
    @     0x7f1260d784e3  _ZSt13__invoke_implIvZN7oneflow6ThreadC4ERKNS0_8StreamIdEEUlvE_JEET_St14__invoke_otherOT0_DpOT1_
    @     0x7f1260d78484  _ZSt8__invokeIZN7oneflow6ThreadC4ERKNS0_8StreamIdEEUlvE_JEENSt15__invoke_resultIT_JDpT0_EE4typeEOS7_DpOS8_
    @     0x7f1260d78422  _ZNSt6thread8_InvokerISt5tupleIJZN7oneflow6ThreadC4ERKNS2_8StreamIdEEUlvE_EEE9_M_invokeIJLm0EEEEvSt12_Index_tupleIJXspT_EEE
    @     0x7f1260d783e3  _ZNSt6thread8_InvokerISt5tupleIJZN7oneflow6ThreadC4ERKNS2_8StreamIdEEUlvE_EEEclEv
    @     0x7f1260d783b8  _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN7oneflow6ThreadC4ERKNS3_8StreamIdEEUlvE_EEEEE6_M_runEv
    @     0x7f120b9d6de4  (unknown)
    @     0x7f12b906c609  start_thread
    @     0x7f12b91a8293  clone
Killing subprocess 986513
Killing subprocess 986514


### 补充
torch.arange 文档链接： https://pytorch.org/docs/stable/generated/torch.arange.html
torch.randperm 文档 https://pytorch.org/docs/stable/generated/torch.randperm.html?highlight=randperm

grybd commented:
合并 fix-all_Sbp4ArgNameAndIndex_bug 分支后跑 test_consistent_randperm.py 还会报错，报错信息：
F20220224 23:03:03.118664 1706926 user_kernel.cpp:171] Check failed: parallel_desc_.hierarchy()->NumAxes() == 1 (2 vs. 1) 
*** Check failure stack trace: ***
F20220224 23:03:03.120785 1706914 user_kernel.cpp:171] Check failed: parallel_desc_.hierarchy()->NumAxes() == 1 (2 vs. 1)

跑 test_consistent_arange.py 报错，报错信息
File "/home/fengdaochao/fdz/oneflow/oneflow/core/job/job_build_and_infer_ctx.cpp", line 335, in CheckOpBlobSplitability
    Check failed: (current_shape.At(axis) % parallel_hierarchy->At(i)) == (0) (1 vs 0) op_name: arange-0 blob_name: out_0 cannot split blob by parallel_hierarchy: 2

@strint @wyg1997 问题应该是比较明确的么，目前还不知道怎么去改，有时间麻烦帮忙看看~

wyg1997 commented:
> 合并 fix-all_Sbp4ArgNameAndIndex_bug 分支后跑 test_consistent_randperm.py 还会报错，报错信息： F20220224 23:03:03.118664 1706926 user_kernel.cpp:171] Check failed: parallel_desc_.hierarchy()->NumAxes() == 1 (2 vs. 1) *** Check failure stack trace: *** F20220224 23:03:03.120785 1706914 user_kernel.cpp:171] Check failed: parallel_desc_.hierarchy()->NumAxes() == 1 (2 vs. 1)
> 
> 跑 test_consistent_arange.py 报错，报错信息 File "/home/fengdaochao/fdz/oneflow/oneflow/core/job/job_build_and_infer_ctx.cpp", line 335, in CheckOpBlobSplitability Check failed: (current_shape.At(axis) % parallel_hierarchy->At(i)) == (0) (1 vs 0) op_name: arange-0 blob_name: out_0 cannot split blob by parallel_hierarchy: 2
> 
> @strint @wyg1997 问题应该是比较明确的么，目前还不知道怎么去改，有时间麻烦帮忙看看~

第二个看起来是不能切分导致的；第一个似乎是在哪里又调用了 SbpParallel4ArgNameAndIndex 接口，定位一下应该不难解决。

grybd commented:
目前[可以跑通](https://github.com/Oneflow-Inc/oneflow/pull/7571#issuecomment-1049989880)  描述的两个问题解决了，解决是否合理还待确认，但是现在跑test_consistent_randperm.py 报错，报错信息：
```shell
 File "/home/fengdaochao/fdz/oneflow/python/oneflow/test/modules/test_consistent_randperm.py", line 30, in _test_consistent_randperm
    test_case.assertEqual(x.sbp, sbp)
AssertionError: Tuples differ: (oneflow.sbp.split(axis=0), oneflow.sbp.split(axis=0)) != (oneflow.sbp.broadcast, oneflow.sbp.split(axis=0)) ```

这个报错是指randperm op 目前还未支持 sbp = (B,S(0)) 时的推导么 @wyg1997 @strint

grybd commented:
在处理支持randperm op的 global tensor的一致性时 与遇到两个问题
1、不知道如何获得GetTensorSliceView4ParallelId(parallel_hierarchy, nd_sbp, logical_shape, parallel_id) 工具函数的这几个参数，我最后的commit 提交的代码是想让 randperm kernel 持有一个private 的TensorSliceView 成员，然后在class CpuRandPermKernel 的CreateOpKernelState(  user_op::KernelInitContext* ctx) 成员函数中利用 ctx 获得GetTensorSliceView4ParallelId(parallel_hierarchy, nd_sbp, logical_shape, parallel_id) 参数，但发现会出现编译出错


2、获得TensorSliceView后如何进行数据拷贝，我发现TensorSliceView 的数据成员只有vector<Range> range_vec_; 
是利用slice 工具函数进行切片保存么类似下下面的一段代码，还是会利用到像tensorsliceviewcopier 类似的拷贝函数，这块我的认知还有些模糊

1) randperm kernel 未改动之前的代码：
```cpp

class CpuRandPermKernel final : public user_op::OpKernel {
 public:
  CpuRandPermKernel() = default;
  ~CpuRandPermKernel() = default;
  std::shared_ptr<user_op::OpKernelState> CreateOpKernelState(
      user_op::KernelInitContext* ctx) const override {
    const auto& generator = CHECK_JUST(one::MakeGenerator(kCPU));
    generator->set_current_seed(ctx->Attr<int64_t>("seed"));
    return std::make_shared<DistributionKernelState>(generator);
  }

 private:
  void Compute(user_op::KernelComputeContext* ctx, user_op::OpKernelState* state,
               const user_op::OpKernelCache*) const override {
    user_op::Tensor* out = ctx->Tensor4ArgNameAndIndex("out", 0);
    int32_t* output = out->mut_dptr<int32_t>();
    const int32_t n = ctx->Attr<int32_t>("n");
    if (n == 0) { return; }
    auto* distribution_state = dynamic_cast<DistributionKernelState*>(state);
    CHECK_NOTNULL(distribution_state);
    const auto& generator = distribution_state->generator();
    const auto& cpu_generator = CHECK_JUST(generator->Get<one::CPUGeneratorImpl>());
    CHECK_NOTNULL(generator);
    user_op::ArangeFunctor<DeviceType::kCPU, int32_t>()(ctx->stream(), 0, 1, n, output);
    std::shuffle(output, output + n, cpu_generator->engine());
  }
  bool AlwaysComputeWhenAllOutputsEmpty() const override { return false; }
};
```
2) 改动后的randperm kernel 的代码

```cpp

class CpuRandPermKernel final : public user_op::OpKernel {
 public:
  CpuRandPermKernel() = default;
  ~CpuRandPermKernel() = default;
  TensorSliceView view;
  std::shared_ptr<user_op::OpKernelState> CreateOpKernelState(
      user_op::KernelInitContext* ctx) const override {
    const auto& generator = CHECK_JUST(one::MakeGenerator(kCPU));
    generator->set_current_seed(ctx->Attr<int64_t>("seed"));
     if (ctx->parallel_ctx().parallel_num() > 1) {
      const Shape& parallel_hierarchy = *ctx->parallel_desc().hierarchy();
  const cfg::NdSbp& nd_sbp = ctx->NdSbp4ArgNameAndIndex("out", 0);
  int32_t n = ctx->Attr<int32_t>("n");
  const Shape& logical_shape = Shape({n});
  const int64_t parallel_id = ctx->parallel_ctx().parallel_id();
   view =
      GetTensorSliceView4ParallelId(parallel_hierarchy, nd_sbp, logical_shape, parallel_id);
   }
    return std::make_shared<DistributionKernelState>(generator);
  }

 private:
  void Compute(user_op::KernelComputeContext* ctx, user_op::OpKernelState* state,
               const user_op::OpKernelCache*) const override {
    user_op::Tensor* out = ctx->Tensor4ArgNameAndIndex("out", 0);
    int32_t* output = out->mut_dptr<int32_t>();
    const int32_t n = ctx->Attr<int32_t>("n");
    int64_t rank_id = GlobalProcessCtx::Rank();
    if (n == 0) { return; }
    auto* distribution_state = dynamic_cast<DistributionKernelState*>(state);
    CHECK_NOTNULL(distribution_state);
    const auto& generator = distribution_state->generator();
    const auto& cpu_generator = CHECK_JUST(generator->Get<one::CPUGeneratorImpl>());
    CHECK_NOTNULL(generator);
    user_op::ArangeFunctor<DeviceType::kCPU, int32_t>()(ctx->stream(), 0, 1, n, output);
    std::shuffle(output, output + n, cpu_generator->engine());
    const std::vector<Range>& range_vec = view.range_vec();
    std::vector<int64_t> start;
    std::vector<int64_t> stop;
    std::vector<int64_t> step(range_vec.size(), 1);
    for (const auto& range : range_vec) {
        start.emplace_back(range.begin());
        stop.emplace_back(range.end());
      }
    output = one::functional::detail::Slice(output, start, stop, step);
    }
  bool AlwaysComputeWhenAllOutputsEmpty() const override { return false; }
  private:
      TensorSliceView view;
};
```   
@strint  麻烦有时间帮看看~

grybd commented:
提交的最新一版代码主要是randperm kernel 绑定 SetInferTmpSizeFn ，使得kernel 可以在 compute里使用临时空间，目前还是遇到了问题，就是获得本rank 上 range 后将临时空间上的数据搬运到 local tensor 上 发现输出的global tensor 不对，可以看到 rank1上的输出信息如下：
```shell
0 2 1 5 9 8 4 7 6 3 84763-----------------------------------------------------------
after S0 global tensor: 
 tensor([0, 2, 1, 5, 9, 0, 0, 0, 0, 0], placement=oneflow.placement(type="cpu", ranks=[0, 1]),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
after S0 local tensor: 
  tensor([0, 0, 0, 0, 0], dtype=oneflow.int64)
```
正确的global tensor输出应该是[0, 2, 1, 5, 9, 8, 4, 7, 6, 3]   @strint @wyg1997

wyg1997 commented on oneflow/user/kernels/randperm_kernel.cpp:
```diff
@@ -39,20 +53,29 @@ class CpuRandPermKernel final : public user_op::OpKernel {
-    std::shuffle(output, output + n, cpu_generator->engine());
+    user_op::ArangeFunctor<DeviceType::kCPU, int32_t>()(ctx->stream(), 0, 1, n, temp);
+    std::shuffle(temp, temp + n, cpu_generator->engine());
+    for (int i = view.At(0).begin(); i < view.At(0).end(); i++) { *(output + i) = *(temp + i); }
```
这里写错了吧，应该是把 `[*(temp+view.At(0)), *(temp+view.At(1)))` 的数据拷贝到 `[*(output), *(output+view.shape().At(0)))`

wyg1997 replied:
其实 debug 很简单的，你写个 for 循环把数据打出来，该怎么对应着拷贝就一目了然嘛

grybd replied:
好的，我看一下~

doombeaker commented:
这个 PR 离 approve 还有多远的距离呢？计划的是周五前得到 approve，我理解最好今天可以开始 review。
但是过去的两天其实没有新的 commit，也没有开始 review。 @grybd

grybd commented:
> 这个 PR 离 approve 还有多远的距离呢？计划的是周五前得到 approve，我理解最好今天可以开始 review。 但是过去的两天其实没有新的 commit，也没有开始 review。 @grybd

目前randperm kernel 已经支持global tensor ,周一提PR解决的，现在处理arange kernel 支持global tensor ,已经有思路了并实施，今天应该可以解决 @doombeaker

doombeaker commented:
> 目前randperm kernel 已经支持global tensor ,周一提PR解决的

这里的 “PR” 指的是这个 commit 吗 https://github.com/Oneflow-Inc/oneflow/pull/7571/commits/1c2be4fdf398200616653eb4ba91f6fd9e7bdd74 ？

我没有看到额外的 PR。

这个周一的 commit，把空行算上，一共5行代码。是周一、周二的主要工作产出吗

grybd commented:
> > 目前randperm kernel 已经支持global tensor ,周一提PR解决的
> 
> 这里的 “PR” 指的是这个 commit 吗 [1c2be4f](https://github.com/Oneflow-Inc/oneflow/commit/1c2be4fdf398200616653eb4ba91f6fd9e7bdd74) ？
> 
> 我没有看到额外的 PR。
> 
> 这个周一的 commit，把空行算上，一共5行代码。是周一、周二的主要工作产出吗。 @grybd

是commit的,我打错了，不好意思，周一，是一直在调试debug前面修改的randperm kernel 支持global tensor 的代码的， arange kernel 支持global tensor的方式和randperm kernel 支持 arange kernel 的方式不一样，所以我查看了一下oneflow 里 如何在kernel里使用cache 解决这些问题的，所以周二我花了大半部分时间在弄懂这些代码逻辑上了，今天主要就是把昨天学的知识应用到解决arange kernel上 ，我尽快解决吧 @doombeaker

strint commented:
这个 @grybd 先整理下思路，看看有什么问题，下午约我和 @wyg1997  再总结下实施方案吧

grybd commented:
> 这个 @grybd 先整理下思路，看看有什么问题，下午约我和 @wyg1997 再总结下实施方案吧

好的，刚看到信息，目前randperm kernel已经支持global tensor ，但还没优化，就是之前yinggang说的数组成片copy,现在要解决的是arange kernel 支持global tensor 的问题，arange kernel 和 randperm kernel 不一样，arange kernel 没有 CreateOpKernelState(user_op::KernelInitContext* ctx) const override {} 函数所以没办法像randperm kernel 那样获得 GetTensorSliceView4ParallelId(hierarchy, nd_sbp, logical_shape, parallel_id);这个函数的几个参数，所以我想着给arange kernel 加一个cache 让每个rank 上缓存 range 的 lower_、upper_；并在arange kernel 中利用 InitOpKernelCache(user_op::KernelCacheContext* ctx) const override 函数进行初始化，这样就可以在compute 中使用本rank 上缓存的 range 的 lower_、upper_,就可以进行将临时空间上的tensor 数据copy到local tensor了，但是修改完代码发现编译出错，报错信息：
```shell
  oneflow::user_op::GenFwInferTmpSizeFn() [with T = double; oneflow::user_op::InferTmpSizeFn = std::function<long unsigned int(oneflow::user_op::InferContext*)>]::<lambda(oneflow::user_op::InferContext*)>
```
有空一起看看代码嘛@strint  @wyg1997

doombeaker commented on oneflow/user/kernels/randperm_kernel.cpp:
```diff
@@ -63,11 +63,8 @@ class CpuRandPermKernel final : public user_op::OpKernel {
-      j++;
-    }
+    auto len = view.At(0).end()-view.At(0).begin();
+    memcpy(output, temp+view.At(0).begin(), sizeof(int32_t)*len);
```
是不是用 `std::copy` 更好。

doombeaker commented on oneflow/user/kernels/randperm_kernel.cpp:
```diff
@@ -39,20 +53,30 @@ class CpuRandPermKernel final : public user_op::OpKernel {
+    user_op::ArangeFunctor<DeviceType::kCPU, int32_t>()(ctx->stream(), 0, 1, n, temp);
+    std::shuffle(temp, temp + n, cpu_generator->engine());
+    auto len = view.At(0).end() - view.At(0).begin();
+    memcpy(output, temp + view.At(0).begin(), sizeof(int32_t) * len);
```
是不是用 `std::copy` 更好一点

grybd replied:
我查一下

doombeaker commented:
这个一个月前开始的 PR，按照约定，应该是 3 月 25 日之前得到 approve。
最近一次的阻碍是编译错误。

查看 commit，编译错误应该已经解决，那这个 PR 离 approve 的距离是什么呢。

grybd commented:
> 这个一个月前开始的 PR，按照约定，应该是 3 月 25 日之前得到 approve。 最近一次的阻碍是编译错误。
> 
> 查看 commit，编译错误应该已经解决，那这个 PR 离 approve 的距离是什么呢。 @grybd
目前arange kernel 已经支持cpu ，cuda端的global tensor 但是目前randperm kernel 只支持了cpu端的global tensor ，cuda 端的还未支持，我目前在修改randperm cuda kernel 使得支持global tensor ；我对cuda这块不太熟悉所以这过程遇到了些问题，zzk叫我把问题描述一下，他有空帮看看，待会会发一个comment描述一下问题

grybd commented:
如何修改randperm cuda kernel使得支持global tensor？

首先这个PR的目的是修改arange kernel 、randperm kernel 使得支持global tansor ,具体方法是-在处理 randperm op 和 arange op 支持 B/S 时 保持 global tensor 的一致性，目前打算采用的处理方案是让多个rank 公用seed 然后在 先在每个rank上生成完整的tensor再根据 infer physic shape信息利GetTensorSliceView4ParallelId(parallel_hierarchy, nd_sbp, logical_shape, parallel_id) 这个工具函数，获得本rank_id 和 physic shape所对应的tensor 上的索引信息(range)，再把对应的位置的数据拷贝到 本rank 的local tensor 上，目前已经修改了cpu 端的randperm kernel 过了我写的测试例子，可以满足global tensor,但是由于对cuda编程体系不太熟悉，目前尚未解决randperm cuda kernel 支持 global tensor的问题，现在就想达到的目的以及这个过程遇到的困惑的点描述一下zzk有空帮看下~
1、如何在randperm 中申请临时空间并使用，我发现randperm cuda kernel的实现与cpu端的实现不同，cpu端主要通过调用 ArangeFunctor先产生随机数，再通过shuffle到各个rank上，所以当我注册的进去的临时空间就可以能直接从这块地址的首地址处进行数据的成片copy，但是cuda 端的randperm kernel 是利用 GeneKeysAndValues函数根据传进行的state生成随机的key和value,然后再通过cub::DeviceRadixSort::SortPairs函数进行按key进行升序排列，并取出其对应的value作为output输出，因为cub::DeviceRadixSort::SortPairs函数需要临时的空间进行排序，所以一开始就需要把用到的临时空间的大小注册进去了，这给我造成了困惑，因为现在需要额外增加一个临时空间用来存储同一个seed生成的global tensor以便于将本rank 上对应range内的tensor copy到local tensor上，我想到把注册进去的临时空间调大，但是在实际使用时这部分内存空间还涉及到内存对齐等操作，使得我不知道最终保存的global tensor在临时空间上的地址，就没办法进行准确的数据拷贝。

2 、在cuda 中如何高效的进行数据拷贝呢，针对目前这种情况

3、 为什么我在randperm cuda kernel 中的compute函数中写 printf("output: %d",output[0]) 会报segment fault 的错呢，我是想通过printf弄清楚tensor 在cuda 上的布局 
@MARD1NO @doombeaker

MARD1NO commented:
> 如何修改randperm cuda kernel使得支持global tensor？
> 
> 首先这个PR的目的是修改arange kernel 、randperm kernel 使得支持global tansor ,具体方法是-在处理 randperm op 和 arange op 支持 B/S 时 保持 global tensor 的一致性，目前打算采用的处理方案是让多个rank 公用seed 然后在 先在每个rank上生成完整的tensor再根据 infer physic shape信息利GetTensorSliceView4ParallelId(parallel_hierarchy, nd_sbp, logical_shape, parallel_id) 这个工具函数，获得本rank_id 和 physic shape所对应的tensor 上的索引信息(range)，再把对应的位置的数据拷贝到 本rank 的local tensor 上，目前已经修改了cpu 端的randperm kernel 过了我写的测试例子，可以满足global tensor,但是由于对cuda编程体系不太熟悉，目前尚未解决randperm cuda kernel 支持 global tensor的问题，现在就想达到的目的以及这个过程遇到的困惑的点描述一下zzk有空帮看下~ 1、如何在randperm cuda kernel中申请临时空间并使用，我发现randperm cuda kernel的实现与cpu端的实现不同，cpu端主要通过调用 ArangeFunctor先产生随机数，再通过shuffle到各个rank上，所以当我注册的进去的临时空间就可以能直接从这块地址的首地址处进行数据的成片copy，但是cuda 端的randperm kernel 是利用 GeneKeysAndValues函数根据传进行的state生成随机的key和value,然后再通过cub::DeviceRadixSort::SortPairs函数进行按key进行升序排列，并取出其对应的value作为output输出，因为cub::DeviceRadixSort::SortPairs函数需要临时的空间进行排序，所以一开始就需要把用到的临时空间的大小注册进去了，这给我造成了困惑，因为现在需要额外增加一个临时空间用来存储同一个seed生成的global tensor以便于将本rank 上对应range内的tensor copy到local tensor上，我想到把注册进去的临时空间调大，但是在实际使用时这部分内存空间还涉及到cuda 内存对齐等操作，使得我不知道最终保存的global tensor在临时空间上的地址，就没办法进行准确的数据拷贝
> 
> 2 、在cuda 中如何高效的进行数据拷贝呢，针对目前这种情况
> 
> 3、 为什么我在randperm cuda kernel 中的compute函数内写 printf("output: %d",output[0]) 会报segment fault 的错呢，可以确定output首地址上是有数据的；我是想通过printf弄清楚tensor 在cuda 上的布局 @MARD1NO @doombeaker

1. tmpbuffersize 加上你所需的额外空间即可。cuda内存对齐只跟显存分配有关，GetCudaAlignedSize是往512取整，比如你tensor大小是32，他还是会会分配512大小，但你只需要根据实际tensor大小拷贝就行了吧，不用管这个对齐。索引下一部分tmpsize才需要关注这个对齐
2. 先不考虑高不高效吧，用CUDA_1D_Kernel_LOOP先拷着先，这个数据量也不大，就是对0-n-1随机打乱
3. 一个是内存地址，一个是gpu显存地址，所以不能打印出来

grybd commented:
> > 如何修改randperm cuda kernel使得支持global tensor？
> > 首先这个PR的目的是修改arange kernel 、randperm kernel 使得支持global tansor ,具体方法是-在处理 randperm op 和 arange op 支持 B/S 时 保持 global tensor 的一致性，目前打算采用的处理方案是让多个rank 公用seed 然后在 先在每个rank上生成完整的tensor再根据 infer physic shape信息利GetTensorSliceView4ParallelId(parallel_hierarchy, nd_sbp, logical_shape, parallel_id) 这个工具函数，获得本rank_id 和 physic shape所对应的tensor 上的索引信息(range)，再把对应的位置的数据拷贝到 本rank 的local tensor 上，目前已经修改了cpu 端的randperm kernel 过了我写的测试例子，可以满足global tensor,但是由于对cuda编程体系不太熟悉，目前尚未解决randperm cuda kernel 支持 global tensor的问题，现在就想达到的目的以及这个过程遇到的困惑的点描述一下zzk有空帮看下~ 1、如何在randperm cuda kernel中申请临时空间并使用，我发现randperm cuda kernel的实现与cpu端的实现不同，cpu端主要通过调用 ArangeFunctor先产生随机数，再通过shuffle到各个rank上，所以当我注册的进去的临时空间就可以能直接从这块地址的首地址处进行数据的成片copy，但是cuda 端的randperm kernel 是利用 GeneKeysAndValues函数根据传进行的state生成随机的key和value,然后再通过cub::DeviceRadixSort::SortPairs函数进行按key进行升序排列，并取出其对应的value作为output输出，因为cub::DeviceRadixSort::SortPairs函数需要临时的空间进行排序，所以一开始就需要把用到的临时空间的大小注册进去了，这给我造成了困惑，因为现在需要额外增加一个临时空间用来存储同一个seed生成的global tensor以便于将本rank 上对应range内的tensor copy到local tensor上，我想到把注册进去的临时空间调大，但是在实际使用时这部分内存空间还涉及到cuda 内存对齐等操作，使得我不知道最终保存的global tensor在临时空间上的地址，就没办法进行准确的数据拷贝
> > 2 、在cuda 中如何高效的进行数据拷贝呢，针对目前这种情况
> > 3、 为什么我在randperm cuda kernel 中的compute函数内写 printf("output: %d",output[0]) 会报segment fault 的错呢，可以确定output首地址上是有数据的；我是想通过printf弄清楚tensor 在cuda 上的布局 @MARD1NO @doombeaker
> 
> 1. tmpbuffersize 加上你所需的额外空间即可。cuda内存对齐只跟显存分配有关，GetCudaAlignedSize是往512取整，比如你tensor大小是32，他还是会会分配512大小，但你只需要根据实际tensor大小拷贝就行了吧，不用管这个对齐。索引下一部分tmpsize才需要关注这个对齐
> 2. 先不考虑高不高效吧，用CUDA_1D_Kernel_LOOP先拷着先，这个数据量也不大，就是对0-n-1随机打乱
> 3. 一个是内存地址，一个是gpu显存地址，所以不能打印出来

好的，谢谢解答~

grybd commented:
目前 randperm kernel 和 arange kernel 支持global tensor 可以通过简单的测试用例,但跑test_consistent_arange.py 和 test_consistent_randperm.py报错，具体见下面几点说明：
1、arange kernel 支持 global tensor的情况
    1）cpu 端支持global tensor的例子：
          ```python
             import numpy as np
              import oneflow as flow
              
              
              def _test_consistent_rand_graph(N,placement):
                class ConsistentRandGraph(flow.nn.Graph):
                    def __init__(self,):
                        super().__init__()
              
                    def build(self):
                        x = flow.arange(1,N,1, placement=placement,sbp=flow.sbp.split(0))
                        return x
              
                c_r_g = ConsistentRandGraph()
                res = c_r_g()
                print("***********************************************************")
                print("after S0 arange kernel global tensor: \n",res)
                print("after S0 arange kernel local tensor: \n ",res.to_local())
              
              if __name__ == "__main__":
                _test_consistent_rand_graph(10,placement=flow.env.all_device_placement("cpu"))
  ```
T0 输出：
```shell
after S0 randperm kernel global tensor: 
 tensor([10,  4,  6,  0,  5,  1,  9,  3,  7,  8,  2],
       placement=oneflow.placement(type="cpu", ranks=[0, 1]), sbp=(oneflow.sbp.split(axis=0),),
       dtype=oneflow.int64)
after S0 randperm kernel local tensor: 
  tensor([10,  4,  6,  0,  5,  1], dtype=oneflow.int64)
```
T1输出：
```shell
after S0 randperm kernel global tensor: 
 tensor([10,  4,  6,  0,  5,  1,  9,  3,  7,  8,  2],
       placement=oneflow.placement(type="cpu", ranks=[0, 1]), sbp=(oneflow.sbp.split(axis=0),),
       dtype=oneflow.int64)
after S0 randperm kernel local tensor: 
  tensor([9, 3, 7, 8, 2], dtype=oneflow.int64)
```
  2）cuda 端支持global tensor的例子：
    T0输出：
     ```shell
     after S0 randperm kernel global tensor: 
 tensor([ 7,  0, 10,  9,  2,  5,  6,  8,  4,  1,  3],
       placement=oneflow.placement(type="cuda", ranks=[0, 1]), sbp=(oneflow.sbp.split(axis=0),),
       dtype=oneflow.int64)
after S0 randperm kernel local tensor: 
  tensor([ 7,  0, 10,  9,  2,  5], device='cuda:0', dtype=oneflow.int64)
```
T1 输出：
```shell
***********************************************************
after S0 arange kernel global tensor: 
 tensor([0, 1, 2, 3, 4, 5, 6, 7, 8], placement=oneflow.placement(type="cuda", ranks=[0, 1]),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
after S0 arange kernel local tensor: 
  tensor([5, 6, 7, 8], device='cuda:1', dtype=oneflow.int64)
```

2、randperm kernel 支持 global tensor的情况
    1）cpu 端支持global tensor的例子：
          ```python
            import numpy as np
           import oneflow as flow
  def _test_consistent_rand_graph(N,placement):
      class ConsistentRandGraph(flow.nn.Graph):
          def __init__(self,):
              super().__init__()
  
          def build(self):
              x = flow.randperm(N, placement=placement,generator=generator, sbp=flow.sbp.split(0))
              return x
  
      c_r_g = ConsistentRandGraph()
      res = c_r_g()
      print("***********************************************************")
      print("after S0 randperm kernel global tensor: \n",res)
      print("after S0 randperm kernel local tensor: \n ",res.to_local())
    
  
  if __name__ == "__main__":
      generator = flow.Generator() 
      generator.manual_seed(0)
      _test_consistent_rand_graph(11,placement=flow.env.all_device_placement("cpu"))

   ```
T0 输出：
```shell
after S0 randperm kernel global tensor: 
 tensor([10,  4,  6,  0,  5,  1,  9,  3,  7,  8,  2],
       placement=oneflow.placement(type="cpu", ranks=[0, 1]), sbp=(oneflow.sbp.split(axis=0),),
       dtype=oneflow.int64)
after S0 randperm kernel local tensor: 
  tensor([10,  4,  6,  0,  5,  1], dtype=oneflow.int64)
```
T1输出：
```shell
after S0 randperm kernel global tensor: 
 tensor([10,  4,  6,  0,  5,  1,  9,  3,  7,  8,  2],
       placement=oneflow.placement(type="cpu", ranks=[0, 1]), sbp=(oneflow.sbp.split(axis=0),),
       dtype=oneflow.int64)
after S0 randperm kernel local tensor: 
  tensor([9, 3, 7, 8, 2], dtype=oneflow.int64)
```
  2）cuda 端支持global tensor的例子：
    T0输出：
     ```shell
        ***********************************************************
 tensor([ 7,  0, 10,  9,  2,  5,  6,  8,  4,  1,  3],
       placement=oneflow.placement(type="cuda", ranks=[0, 1]), sbp=(oneflow.sbp.split(axis=0),),
       dtype=oneflow.int64)
after S0 randperm kernel local tensor: 
  tensor([ 7,  0, 10,  9,  2,  5], device='cuda:0', dtype=oneflow.int64)
```
T1 输出：
```shell
after S0 randperm kernel global tensor: 
 tensor([ 7,  0, 10,  9,  2,  5,  6,  8,  4,  1,  3],
       placement=oneflow.placement(type="cuda", ranks=[0, 1]), sbp=(oneflow.sbp.split(axis=0),),
       dtype=oneflow.int64)
after S0 randperm kernel local tensor: 
  tensor([6, 8, 4, 1, 3], device='cuda:1', dtype=oneflow.int64)
```
但是跑test_consistent_arange.py 和 test_consistent_randperm.py报错，具体看下面2点说明

1、跑test_consistent.arange.py报错，报错信息：
```shell
Segmentation fault (core dumped)
```
2、跑test_consistent_randperm.py报错，报错信息：
```shell
Ran 2 tests in 0.118s

OK (skipped=1)
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)
Aborted (core dumped)
```
但是跑test_consistent_arange.py和test_consistent_randperm.py会出现越界抛出异常的情况，错误信息：
```shell
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)
Aborted (core dumped) 
```
看报错信息，是内存访问越界的问题，但不知道为什么会报错@strint @wyg1997 有空帮忙看看哈~

strint commented:
你这个PR还是存在之前说的问题，想要一次解决4个问题 arange/perm * cpu/gpu。这讨论你写一个issue，一次写4个问题，很难看清楚进展和问题。所以：
- 先把PR拆分成两个，这个PR下解决arange的问题；下个PR再解决randperm的问题。randperm可以先不管。
- 我们只看arrage的问题，arrage 的cpu看起来结果是正常的，但是gpu你看下，结果是异常的，没有“1”

grybd commented:
> 你这个PR还是存在之前说的问题，想要一次解决4个问题 arange/perm * cpu/gpu。这讨论你写一个issue，一次写4个问题，很难看清楚进展和问题。所以：
> 
> * 先把PR拆分成两个，这个PR下解决arange的问题；下个PR再解决randperm的问题。randperm可以先不管。
> * 我们只看arrage的问题，arrage 的cpu看起来结果是正常的，但是gpu你看下，结果是异常的，没有“1”

好的，刚发现测试arange op的输出结果贴错了，写太长不注意把测试 randperm op的结果往上贴覆盖掉了，不好意思，我改过来了现在， 昨晚是重新提了个PR单独解决Arange kernel 支持global tensor ： https://github.com/Oneflow-Inc/oneflow/pull/7909 今天想把问题归结一起说一下，所以混着说了，我待会就提一下issue分别描述一下，这样会比较清晰 @strint

strint commented on oneflow/user/kernels/randperm_kernel.cu:
```diff
@@ -96,29 +144,54 @@ class GpuRandPermKernel final : public user_op::OpKernel {
+          /* d_keys_in */ key_base,
+          /* d_keys_out */ key_base + n,
+          /* d_values_in */ value_base,
+          /* d_values_out */ temp_buffer_base,
```
这里还应该是 output ？

eager local 的单测也本地跑一下，应该过不了吧

strint commented on python/oneflow/test/modules/test_consistent_arange.py:
```diff
@@ -1,71 +0,0 @@
-from oneflow.test_utils.automated_test_util import *
-
-
-@autotest(n=1, auto_backward=False, check_graph=False)
```
这个是误删的？

strint commented on oneflow/user/kernels/randperm_kernel.cu:
```diff
@@ -96,29 +144,54 @@ class GpuRandPermKernel final : public user_op::OpKernel {
   bool AlwaysComputeWhenAllOutputsEmpty() const override { return false; }
+
+ private:
+  mutable TensorSliceView view;
```
看起来不适合作为成员，当临时变量用就好

strint commented on oneflow/user/kernels/randperm_kernel.cpp:
```diff
@@ -35,24 +68,40 @@ class CpuRandPermKernel final : public user_op::OpKernel {
   bool AlwaysComputeWhenAllOutputsEmpty() const override { return false; }
+
+ private:
+  mutable TensorSliceView view;
```
看起来不适合作为成员，当临时变量用就好

strint commented on oneflow/user/kernels/randperm_kernel.cu:
```diff
@@ -59,17 +80,44 @@ class GpuRandPermKernel final : public user_op::OpKernel {
     const auto& generator = CHECK_JUST(one::MakeGenerator(kCUDA));
     generator->set_current_seed(ctx->Attr<int64_t>("seed"));
+
+    int64_t parallel_num = ctx->parallel_ctx().parallel_num();
```
这一段看起来是没用的。因为都是用一个generator

strint commented on oneflow/user/kernels/randperm_kernel.cu:
```diff
@@ -59,17 +80,44 @@ class GpuRandPermKernel final : public user_op::OpKernel {
+      view = GetTensorSliceView4ParallelId(hierarchy, nd_sbp, logical_shape, parallel_id);
+      std::shared_ptr<GpuRandPermKernelCache> cache(
+          new GpuRandPermKernelCache(view.At(0).begin(), view.At(0).end()));
+      return cache;
```
这里的view的起点终点，是应该方cache还是state，我也有疑问

请 @daquexian 帮忙看看吧。

我感觉cache本身和state，从名字上没有看出明显的区分度

daquexian replied:
这里放在 cache 是合理的，cache 的目的是节省重复计算，清空对正确性也没有影响，state 表示 “状态”（比如 ofrecord reader 的 index）

grybd commented:
已经按照建议修改，并在本地可以跑通了testcase ,有空再帮忙看看~ @strint  @wyg1997

wyg1997 commented on python/oneflow/test/modules/test_consistent_randperm.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def _test_consistent_randperm(test_case, N, placement, sbp, dtype):
+    x = flow.randperm(N, placement=placement, sbp=sbp, dtype=dtype)
+    test_case.assertEqual(x.dtype, dtype)
+    test_case.assertEqual(x.sbp, sbp)
+    test_case.assertEqual(x.placement, placement)
```
这个算子的数值比较特殊，还是建议测试一下，具体的做法是：`x.numpy()` 拿到 broadcast 的结果后，走一遍 numpy.sort，然后和 numpy.arange 来对比

wyg1997 replied:
@strint

strint replied:
我验证了一下，eager因为随机数种子没有做全局同步，所以数值还不对齐，shape对齐了。

graph可以数值也对齐。

grybd commented:
好的，我处理一下 @wyg1997 @strint

grybd commented:
已经按照建议添加了数值验证测试 @wyg1997 @strint

wyg1997 commented on python/oneflow/test/modules/test_consistent_randperm.py:
```diff
@@ -0,0 +1,90 @@
+    test_case.assertEqual(x.placement, placement)
+
+
+class TestRandConsistent(flow.unittest.TestCase):
```
这里应该是 TestRandpermConsistent

wyg1997 commented on python/oneflow/test/modules/test_consistent_randperm.py:
```diff
@@ -0,0 +1,90 @@
+
+class TestRandConsistent(flow.unittest.TestCase):
+    @globaltest
+    def test_rand_consistent(test_case):
```
```suggestion
    def test_randperm_consistent(test_case):
```

wyg1997 commented on python/oneflow/test/modules/test_consistent_randperm.py:
```diff
@@ -0,0 +1,90 @@
+                        _test_consistent_randperm(test_case, N, placement, sbp, dtype)
+
+    @globaltest
+    def test_rand_graph(test_case):
```
```suggestion
    def test_randperm_graph(test_case):
```

grybd commented:
[/opt/conda/bin/python3 /opt/conda/lib/python3.7/site-packages/oneflow/nn/modules/random_ops.py -v]     y = flow.randperm(5, generator=generator) # construct local tensor
[/opt/conda/bin/python3 /opt/conda/lib/python3.7/site-packages/oneflow/nn/modules/random_ops.py -v] Expecting nothing
[/opt/conda/bin/python3 /opt/conda/lib/python3.7/site-packages/oneflow/nn/modules/random_ops.py -v] Segmentation fault (core dumped)
[/opt/conda/bin/python3 /opt/conda/lib/python3.7/site-packages/oneflow/nn/modules/random_ops.py -v] 
    return future.result()
  File "/home/ci-user/runners/gpu/_work/oneflow/oneflow/ci/test/parallel_run.py", line 106, in run_cmds
    for cmd_gpu_num in everyN(cmdN, gpu_num)
  File "/home/ci-user/runners/gpu/_work/oneflow/oneflow/ci/test/parallel_run.py", line 88, in spawn_shell_and_check
    raise RuntimeError(cmd)
RuntimeError: /opt/conda/bin/python3 /opt/conda/lib/python3.7/site-packages/oneflow/nn/modules/random_ops.py -v
Error: Process completed with exit code 1.

strint commented:
现在报错是这里：
```
https://github.com/Oneflow-Inc/oneflow/blob/source_op_support_S/python/oneflow/nn/modules/random_ops.py#L432
```
是 doc test。

你用这个分支，本地运行这个case看看 @grybd

grybd commented:
> 现在报错是这里：
> 
> ```
> https://github.com/Oneflow-Inc/oneflow/blob/source_op_support_S/python/oneflow/nn/modules/random_ops.py#L432
> ```
> 
> 是 doc test。
> 
> 你用这个分支，本地运行这个case看看 @grybd
> 
> 当前看起来是：
> 
> ```
> y = flow.randperm(5, generator=generator) # construct local tensor
> ```
> 
> 会挂掉

好的

strint commented on oneflow/user/kernels/randperm_kernel.cpp:
```diff
@@ -20,12 +20,46 @@ limitations under the License.
+  std::shared_ptr<user_op::OpKernelCache> InitOpKernelCache(
+      user_op::KernelCacheContext* ctx) const override {
+    int64_t parallel_num = ctx->parallel_ctx().parallel_num();
+    const NdSbp& nd_sbp = ctx->NdSbp4ArgNameAndIndex("out", 0);
```
这一行估计需要写到 `if (parallel_num > 1) {` 里面

grybd commented:
好的，我改了试试

grybd commented:
FAILED python/oneflow/test/modules/test_one_embedding_ftrl.py::TestOptimizers::test_ftrl
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
==== 1 failed, 854 passed, 180 skipped, 996 warnings in 3232.69s (0:53:52) =====

grybd commented:
本地跑没有问题
